File: ext/inject_schema.py
import json
import os
import re
from datetime import datetime

# ======================================================
# KONFIGURASI GLOBAL (V7.0.0 - Self-Signature Edition)
# ======================================================
BASE_URL = "https://dalam.web.id"
SITE_NAME = "Layar Kosong"
AUTHOR = "Fakhrul Rijal"
LICENSE_URL = "https://creativecommons.org/publicdomain/zero/1.0/"

ALLOWED_CATEGORIES = {
    "gaya-hidup", "jejak-sejarah", "lainnya",
    "olah-media", "opini-sosial", "sistem-terbuka", "warta-tekno"
}

SCHEMA_REGEX = re.compile(
    r'<script\s+type="application/ld\+json">.*?</script>',
    re.DOTALL | re.IGNORECASE
)

# Tanda unik untuk mengecek apakah file sudah diproses
SIGNATURE_KEY = "schema_oleh_Fakhrul_Rijal"

STOPWORDS = {"yang", "untuk", "dengan", "adalah", "dalam", "dari", "pada", "atau", "itu", "dan", "sebuah", "aku", "ke", "saya", "ini", "gue", "gua", "elu", "elo"}

# ======================================================
# UTILITIES
# ======================================================
def slugify(text: str) -> str:
    return text.strip().lower().replace(" ", "-")

def category_name_clean(category: str) -> str:
    return category.strip().replace("-", " ").title()

def build_keywords(headline: str, category: str, slug: str) -> str:
    words = re.split(r"[^\w]+", headline.lower())
    refined_words = [w for w in words if len(w) > 3 and w not in STOPWORDS]
    base = set(refined_words)
    base.add(category.lower())
    slug_parts = slug.replace("-", " ").split()
    base.update([s for s in slug_parts if len(s) > 3])
    base.add("layar kosong")
    return ", ".join(sorted(base)[:12])

# ======================================================
# SCHEMA BUILDER
# ======================================================
def build_combined_schema(category, article):
    headline, filename, image, iso_date, desc = article
    cat_slug = slugify(category)
    file_slug = filename.replace('.html', '').lstrip('/')
    
    clean_base = BASE_URL.rstrip('/')
    article_url = f"{clean_base}/{cat_slug}/{file_slug}".rstrip('/')
    category_url = f"{clean_base}/{cat_slug}".rstrip('/')
    
    cat_display_name = category_name_clean(category)
    keywords = build_keywords(headline, cat_display_name, file_slug)

    schema_data = {
        "@context": "https://schema.org",
        "@graph": [
            {
                "@type": "WebSite",
                "@id": f"{clean_base}#website",
                "url": clean_base,
                "name": SITE_NAME,
                "publisher": {
                    "@type": "Organization",
                    "@id": f"{clean_base}#organization",
                    "name": SITE_NAME,
                    "url": clean_base,
                    "logo": {
                        "@type": "ImageObject",
                        "url": f"{clean_base}/logo.png",
                        "width": 384,
                        "height": 384
                    }
                }
            },
            {
                "@type": "Article",
                "@id": f"{article_url}#article",
                "isPartOf": {"@id": f"{clean_base}#website"},
                "mainEntityOfPage": {"@type": "WebPage", "@id": article_url},
                "license": LICENSE_URL,
                "headline": headline,
                "description": desc,
                "articleSection": cat_display_name,
                "keywords": keywords,
                "image": {"@type": "ImageObject", "url": image, "width": 1200, "height": 675},
                "author": {"@type": "Person", "name": AUTHOR, "url": f"{clean_base}/about"},
                "publisher": {"@id": f"{clean_base}#organization"},
                "datePublished": iso_date,
                "dateModified": iso_date
            },
            {
                "@type": "BreadcrumbList",
                "@id": f"{article_url}#breadcrumb",
                "itemListElement": [
                    {"@type": "ListItem", "position": 1, "name": "Beranda", "item": clean_base},
                    {"@type": "ListItem", "position": 2, "name": cat_display_name, "item": category_url},
                    {"@type": "ListItem", "position": 3, "name": headline, "item": article_url}
                ]
            }
        ]
    }
    return (
        '<script type="application/ld+json">'
        + json.dumps(schema_data, ensure_ascii=False, separators=(",", ":"))
        + "</script>\n"
    )

# ======================================================
# EKSEKUSI UTAMA
# ======================================================
if __name__ == "__main__":
    if not os.path.exists("artikel.json"):
        print("❌ File artikel.json tidak ditemukan!")
        exit()

    with open("artikel.json", "r", encoding="utf-8") as f:
        data = json.load(f)

    results = {"changed": 0, "skipped": 0, "missing": 0, "forbidden": 0}
    tgl = datetime.now().strftime("%Y-%m-%d")
    signature = f"<noscript>{SIGNATURE_KEY}_{tgl}</noscript>"

    for category, articles in data.items():
        cat_slug = slugify(category)

        if cat_slug not in ALLOWED_CATEGORIES:
            results["forbidden"] += len(articles)
            continue

        for article in articles:
            filename = article[1]
            html_path = os.path.join(cat_slug, filename.lstrip('/'))

            if not os.path.isfile(html_path):
                results["missing"] += 1
                continue

            with open(html_path, "r", encoding="utf-8") as f:
                html_content = f.read()

            # === CEK SIGNATURE (Anti-Duplikat Lokal) ===
            if SIGNATURE_KEY in html_content:
                results["skipped"] += 1
                continue

            print(f"🧠 Injecting Schema & Signature: {html_path}")

            # 1. Bersihkan schema lama (jika ada)
            html_clean = re.sub(SCHEMA_REGEX, "", html_content)
            
            # 2. Build schema baru
            inject_code = build_combined_schema(category, article)

            # 3. Sisipkan schema ke bagian Head
            if "<style>" in html_clean:
                new_html = html_clean.replace("<style>", f"{inject_code}<style>")
            elif "</head>" in html_clean:
                new_html = html_clean.replace("</head>", f"{inject_code}</head>")
            else:
                new_html = inject_code + html_clean

            # 4. Tambahkan signature di paling bawah file (seperti script JS kamu)
            final_html = new_html.strip() + "\n" + signature

            with open(html_path, "w", encoding="utf-8") as f:
                f.write(final_html)

            results["changed"] += 1

    print("-" * 50)
    print(f"✅ SEO Schema Injector V7.0.0 (Noscript Signature Mode)")
    print(f"🆕 Di-inject & Terpatri : {results['changed']}")
    print(f"⏭️  Skip (Sudah Ada)    : {results['skipped']}")
    print(f"❌ File Tidak Ada      : {results['missing']}")
    print(f"🚫 Kategori Dilarang   : {results['forbidden']}")
    print("-" * 50)

File: alat/inject-schema.rs
use serde_json::{json, Value};
use std::collections::HashSet;
use std::fs;
use std::path::Path;
use chrono::Local;
use regex::Regex;
use std::collections::HashMap;

// ======================================================
// KONFIGURASI GLOBAL
// ======================================================
const BASE_URL: &str = "https://dalam.web.id";
const SITE_NAME: &str = "Layar Kosong";
const AUTHOR: &str = "Fakhrul Rijal";
const LICENSE_URL: &str = "https://creativecommons.org/publicdomain/zero/1.0/";
const SIGNATURE_KEY: &str = "schema_oleh_Fakhrul_Rijal";

// Stopwords untuk keyword generator
static STOPWORDS: &[&str] = &["yang", "untuk", "dengan", "adalah", "dalam", "dari", "pada", "atau", "itu", "dan", "sebuah", "aku", "ke", "saya", "ini", "gue", "gua", "elu", "elo"];

// ======================================================
// UTILITIES
// ======================================================
fn slugify(text: &str) -> String {
    text.trim().to_lowercase().replace(" ", "-")
}

fn category_name_clean(category: &str) -> String {
    category.trim().replace("-", " ")
        .split_whitespace()
        .map(|w| {
            let mut c = w.chars();
            match c.next() {
                None => String::new(),
                Some(f) => f.to_uppercase().collect::<String>() + c.as_str(),
            }
        })
        .collect::<Vec<_>>()
        .join(" ")
}

fn build_keywords(headline: &str, category: &str, slug: &str) -> String {
    let re = Regex::new(r"[^\w]+").unwrap();
    let words = re.split(headline.to_lowercase().as_str());

    let mut base: HashSet<String> = words
        .filter(|w| w.len() > 3 && !STOPWORDS.contains(w))
        .map(|w| w.to_string())
        .collect();

    base.insert(category.to_lowercase());

    for part in slug.replace("-", " ").split_whitespace() {
        if part.len() > 3 {
            base.insert(part.to_string());
        }
    }
    base.insert("layar kosong".to_string());

    let mut sorted_keys: Vec<_> = base.into_iter().collect();
    sorted_keys.sort();
    sorted_keys.truncate(12);
    sorted_keys.join(", ")
}

// ======================================================
// SCHEMA BUILDER
// ======================================================
fn build_combined_schema(category: &str, article: &[String]) -> String {
    let headline = &article[0];
    let filename = &article[1];
    let image = &article[2];
    let iso_date = &article[3];
    let desc = &article[4];

    let cat_slug = slugify(category);
    let file_slug = filename.replace(".html", "").trim_start_matches('/').to_string();

    let clean_base = BASE_URL.trim_end_matches('/');
    let article_url = format!("{}/{}/{}", clean_base, cat_slug, file_slug);
    let category_url = format!("{}/{}", clean_base, cat_slug);

    let cat_display_name = category_name_clean(category);
    let keywords = build_keywords(headline, &cat_display_name, &file_slug);

    let schema = json!({
        "@context": "https://schema.org",
        "@graph": [
            {
                "@type": "WebSite",
                "@id": format!("{}#website", clean_base),
                "url": clean_base,
                "name": SITE_NAME,
                "publisher": {
                    "@type": "Organization",
                    "@id": format!("{}#organization", clean_base),
                    "name": SITE_NAME,
                    "url": clean_base,
                    "logo": {
                        "@type": "ImageObject",
                        "url": format!("{}/logo.png", clean_base),
                        "width": 384,
                        "height": 384
                    }
                }
            },
            {
                "@type": "Article",
                "@id": format!("{}#article", article_url),
                "isPartOf": { "@id": format!("{}#website", clean_base) },
                "mainEntityOfPage": { "@type": "WebPage", "@id": article_url },
                "license": LICENSE_URL,
                "headline": headline,
                "description": desc,
                "articleSection": cat_display_name,
                "keywords": keywords,
                "image": { "@type": "ImageObject", "url": image, "width": 1200, "height": 675 },
                "author": { "@type": "Person", "name": AUTHOR, "url": format!("{}/about", clean_base) },
                "publisher": { "@id": format!("{}#organization", clean_base) },
                "datePublished": iso_date,
                "dateModified": iso_date
            },
            {
                "@type": "BreadcrumbList",
                "@id": format!("{}#breadcrumb", article_url),
                "itemListElement": [
                    { "@type": "ListItem", "position": 1, "name": "Beranda", "item": clean_base },
                    { "@type": "ListItem", "position": 2, "name": cat_display_name, "item": category_url },
                    { "@type": "ListItem", "position": 3, "name": headline, "item": article_url }
                ]
            }
        ]
    });

    format!("<script type=\"application/ld+json\">{}</script>\n", serde_json::to_string(&schema).unwrap())
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let allowed_categories: HashSet<&str> = ["gaya-hidup", "jejak-sejarah", "lainnya", "olah-media", "opini-sosial", "sistem-terbuka", "warta-tekno"].iter().cloned().collect();

    let data_str = fs::read_to_string("artikel.json")?;
    let data: HashMap<String, Vec<Vec<String>>> = serde_json::from_str(&data_str)?;

    let mut changed = 0;
    let mut skipped = 0;
    let mut missing = 0;

    let tgl = Local::now().format("%Y-%m-%d").to_string();
    let signature = format!("<noscript>{}_{}</noscript>", SIGNATURE_KEY, tgl);
    let re_schema = Regex::new(r"(?is)<script\s+type=[\x22\x27]application/ld\+json[\x22\x27]>.*?</script>").unwrap();

    for (category, articles) in data {
        let cat_slug = slugify(&category);
        if !allowed_categories.contains(cat_slug.as_str()) { continue; }

        for article in articles {
            let filename = &article[1];
            let html_path = Path::new(&cat_slug).join(filename.trim_start_matches('/'));

            if !html_path.exists() {
                missing += 1;
                continue;
            }

            let mut html_content = fs::read_to_string(&html_path)?;

            if html_content.contains(SIGNATURE_KEY) {
                skipped += 1;
                continue;
            }

            println!("🧠 Injecting Schema: {:?}", html_path);

            // 1. Bersihkan schema lama
            html_content = re_schema.replace_all(&html_content, "").to_string();

            // 2. Build schema baru
            let inject_code = build_combined_schema(&category, &article);

            // 3. Sisipkan ke Head
            let mut new_html = if html_content.contains("<style>") {
                html_content.replace("<style>", &format!("{}<style>", inject_code))
            } else if html_content.contains("</head>") {
                html_content.replace("</head>", &format!("{}</head>", inject_code))
            } else {
                format!("{}{}", inject_code, html_content)
            };

            // 4. Tambahkan signature
            new_html = format!("{}\n{}", new_html.trim(), signature);

            fs::write(&html_path, new_html)?;
            changed += 1;
        }
    }

    println!("✅ SEO Schema Injector Selesai");
    println!("🆕 Di-inject: {}, ⏭️ Skipped: {}, ❌ Missing: {}", changed, skipped, missing);

    Ok(())
}
